### ISSUE TITLE
BUG: Scaling of diff_step if scipy.optimize.least_squares uses 'x' instead of 'x_scale'

### ISSUE THREAD

**Issue by squirion** [scipy/scipy#18161]
### Describe your issue.

In the API reference for the `diff_step` argument in `scipy.optimize.least_squares`, it mentions that 
> The actual step is computed as x * diff_step.

and refers to `William H. Press et. al., “Numerical Recipes. The Art of Scientific Computing. 3rd edition”, Sec. 5.7.`

But the text in Numerical Recipes clearly mentions that the diff_step should be scaled according to `a characteristic scale x_c`, which scale can be assumed to be equal to `x` in the absence of any other information.

In our case, if we provide an argument `x_scale` to the function call, then the actual step  should be computed as:
`x_scale * diff_step`.

As noted in the reference, this is especially important near x=0:
> In the absence of any other information, one often assumes `x_c = x` (except near `x=0`, where some other estimate of the typical `x` scale should be used).

otherwise we lose any control over the effective differentiation step in that region of the search space.

The reproducing code shows that the effective step is in fact not `x_scale*diff_step` as suggested in the reference.

### Reproducing Code Example

```python
def test_effective_diff_step():
    import scipy.optimize as opt

    def absolute_value_fct(x):
        print('Objective function tested at x=',x)
        return abs(x)

    params = {
        'fun': absolute_value_fct,
        'x0': [0],
        'diff_step': [0.1],
        'x_scale': [1.0],
    }
    opt.least_squares(**params)

test_effective_diff_step()
```

### Error message

```shell
Objective function tested at x= [0.]
Objective function tested at x= [1.49011612e-08]
```

### SciPy/NumPy/Python version and system information

```shell
1.10.1 1.24.2 sys.version_info(major=3, minor=8, micro=0, releaselevel='final', serial=0)
C:\Python\Python-3.8\lib\site-packages\scipy\__config__.py:140: UserWarning: Install `pyyaml` for better output
  warnings.warn("Install `pyyaml` for better output", stacklevel=1)
{
  "Compilers": {
    "c": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "10.3.0",
      "commands": "cc"
    },
    "cython": {
      "name": "cython",
      "linker": "cython",
      "version": "0.29.33",
      "commands": "cython"
    },
    "c++": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "10.3.0",
      "commands": "c++"
    },
    "fortran": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "10.3.0",
      "commands": "gfortran"
    },
    "pythran": {
      "version": "0.12.1",
      "include directory": "C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pip-build-env-yajwq4a4\\overlay\\Lib\\site-packages\\pythran"
    }
  },
  "Machine Information": {
    "host": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "windows"
    },
    "build": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "windows"
    },
    "cross-compiled": false
  },
  "Build Dependencies": {
    "blas": {
      "name": "openblas",
      "found": true,
      "version": "0.3.18",
      "detection method": "pkgconfig",
      "include directory": "c:/opt/openblas/if_32/64/include",
      "lib directory": "c:/opt/openblas/if_32/64/lib",
      "openblas configuration": "USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS= NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= PRESCOTT MAX_THREADS=4",
      "pc file directory": "c:/opt/openblas/if_32/64/lib/pkgconfig"
    },
    "lapack": {
      "name": "openblas",
      "found": true,
      "version": "0.3.18",
      "detection method": "pkgconfig",
      "include directory": "c:/opt/openblas/if_32/64/include",
      "lib directory": "c:/opt/openblas/if_32/64/lib",
      "openblas configuration": "USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS= NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= PRESCOTT MAX_THREADS=4",
      "pc file directory": "c:/opt/openblas/if_32/64/lib/pkgconfig"
    }
  },
  "Python Information": {
    "path": "C:\\Users\\runneradmin\\AppData\\Local\\Temp\\cibw-run-5o1gjg1j\\cp38-win_amd64\\build\\venv\\Scripts\\python.exe",
    "version": "3.8"
  }
}
```

**Comment by tupui** [comment]
Hi @squirion, thank you for reporting. I would just like to point out that although we cite "Numerical Recipes", we do not strictly implement what is described. This is for multiple reasons, including licensing/copyrighting.

**Comment by tupui** [comment]
I agree 👍 I am not familiar with that part though so will leave others chime in 😄

**Comment by squirion** [comment]
I dug a little deeper in the code and found that the effective step is computed using the function `_compute_absolute_step(...)` in `scipy\optimize\_numdiff.py`:

```
def _compute_absolute_step(rel_step, x0, f0, method):
    # this is used instead of np.sign(x0) because we need
    # sign_x0 to be 1 when x0 == 0.
    sign_x0 = (x0 >= 0).astype(float) * 2 - 1

    rstep = _eps_for_method(x0.dtype, f0.dtype, method)

    if rel_step is None:
        abs_step = rstep * sign_x0 * np.maximum(1.0, np.abs(x0))
    else:
        # User has requested specific relative steps.
        # Don't multiply by max(1, abs(x0) because if x0 < 1 then their
        # requested step is not used.
        abs_step = rel_step * sign_x0 * np.abs(x0)

        # however we don't want an abs_step of 0, which can happen if
        # rel_step is 0, or x0 is 0. Instead, substitute a realistic step
        dx = ((x0 + abs_step) - x0)
        abs_step = np.where(dx == 0,
                            rstep * sign_x0 * np.maximum(1.0, np.abs(x0)),
                            abs_step)

    return abs_step
```

With Git Blame, we can see that a little while ago (~4yr), that function was simply:
```
def _compute_absolute_step(rel_step, x0, f0, method):
    if rel_step is None:
        rel_step = _eps_for_method(x0.dtype, f0.dtype,method)
    sign_x0 = (x0 >= 0).astype(float) * 2 - 1
    return rel_step * sign_x0 * np.maximum(1.0, np.abs(x0))
```

So, in that earlier revision of the code, when we had `x<1.0`, we simply multiplied the step by `1.0`, which would avoid the obvious issues around `x=0`.

Thus, I would be tempted to suggest replacing the line 181 from `abs_step = rel_step * sign_x0 * np.abs(x0)` in the current code by `abs_step = rel_step * sign_x0 * np.maximum(1.0, np.abs(x0))`... if it weren't for the fact that the comment above specifically says 
> Don't multiply by max(1, abs(x0) because if x0 < 1 then their requested step is not used.

...although that warning doesn't really make sense to me, because the best way to consider the requested step is to multiply by `max(1, abs(x0))` 🤔

**Comment by andyfaff** [comment]
If `rel_step` is not specified then `_compute_absolute_step` tries to figure out a sensible relative step, and takes into account any entries in `x0` that are close to zero, by virtue of `np.maximum(1.0, np.abs(x0))`.

However, if `rel_step` is specified then `_compute_absolute_step` respects the users relative step with respect to `x`. i.e. when the docstring for `least_squares.diff_step` says:

> The actual step is computed as ``x * diff_step``

that is exactly what happens in the numerical differentiation code.  You can change the size of the relative step for each of the entries in `x` by providing an array for `diff_step`. There is also some safety provided for when entries in `x` are zero.

I don't have a copy of Numerical Recipes so can't comment on what the text says in the reference.

Note that if we calculated step size as `x_scale * diff_step` then the steps would become absolute and not relative, which is not desired.

 There's more background in https://github.com/scipy/scipy/issues/11258 as to why we do this.

**Comment by squirion** [comment]
@andyfaff you are right that using `x_scale * diff_step` would make it an absolute step and not a relative one. At first I didn't grasp the importance of the step being relative (I could rephrase my original post if it helps).

Also, you are right when you say that the documentation explains correctly what the code actually does (which is `x * diff_step`)

I read the thread in #11258 and it helps me better understand the big picture, so let me try to reformulate what I see as the remaining problem given this context: while the current algorithm does scale the step with the value of `x` (which from my current understanding is desired at least for numerical reasons), and does provide safety when x=0, it does not account for what I would call "the resolution of the cost function" for lack of a better term.

When working with noisy data, steps smaller than a certain delta will provide meaningless gradients because at that scale the topology of the cost function is dominated by the noise more than the signal. Or, in my case, I have cost functions that have tons and tons of tiny plateaux. If you use a step of `eps` in those functions, the gradients are null, but if I can make sure the step is at least 0.000001 for example, then the function will optimize nicely.

So I think what's really missing there is a way to specify an absolute minimum for the step size.

Using `np.maximum(1.0, np.abs(x0))` has that effect for me, but I understand it has adverse effects as well.

**Comment by squirion** [comment]
Regarding Numerical Recipes, without going into the details, it mentions that the step should be scaled by the "characteristic scale of x", which they dub x_c. 

They say:
> [...] one often assumes x_c = x (except near x=0, where some other estimate of the typical x scale should be used).

which I understand to mean that around `x=0`, special care must be taken to ensure that the step still allows a valid estimation of the gradient.

**Comment by squirion** [comment]
Also, in my specific case, another solution would be to be able to pass an absolute diff_step to the function `least_squares`, but that does not seem to be possible at the moment.

Thanks for the help!
It's great to see such a reactive community!

**Comment by andyfaff** [comment]
Note: the reference to Numerical Recipes is describing numerical differentiation, not topics related to least squares.

I think the current `diff_step` behaviour is better suited to differentiating w.r.t small valued parameters whose partial derivatives are less noisy/have no plateaus. This is because `_compute_absolute_step` allows small valued parameters (`x < 1`) to retain a relative step. If it were changed back to multiplying by `np.maximum(1, np.abx(x0))` then small-valued parameters are given an absolute step. This isn't suited for parameters close to 1.

--> existing behaviour is consistent with NR because we're scaling steps w.r.t the characteristic scale of x. This is True for parameters greater than, and less than, one. For parameters less than 1 that scaling is done by choosing appropriate `diff_step`. 

So what to do with parameters that have plateaus in $\frac{\delta f}{\delta x_i}$? One way might be to create a minimum `abs_step` keyword parameter. However, I don't understand why knowledge of the problem can't be used in this particular issue. You know that there are plateaus, and you have an inkling of what an absolute step should be.

e.g. Let's say that you know that the scale of a parameter might be on the order of 1e-4, but you know steps have to be greater than 2e-6 because of plateaus in $\frac{\delta f}{\delta x_i}$. For that parameter an appropriate `diff_step` might be 0.05 (0.1?). This would make the step 5e-6. This would give you a bit of range that the parameter value could take.

Perhaps a better approach is to transform the parameter itself. If there's a parameter that is known to be small in the objective function why not change the setup to vary the logarithm of the parameter?
e.g. instead of the parameter value being 1e-4 it's now transformed to -4. This does mean that the parameter is forced to be positive, but now it can approach as close to zero as you want.

**Comment by squirion** [comment]
An example might help convey the issue:

Imagine we want to find the x,y coordinates of a white circle on a black background in a pixelated image by comparing the image to a (pixelated) render made from a perfect circle.
Consider that we know the distribution of the possible position for the circle, so that we call the mean position (x0,x1) = (0.0,0.0).
Now, to find that position, you render the circle at various positions, ranging from (-10.0,-10.0) to (10.0,10.0) for example, using a cost function that compares the render to the reference image. We navigate that search space using the least_squares algorithm to find the position that matches the reference image the best.

There are two important aspects in that (much simplified) version of my problem:
1. We always start at x=0 as it is our best guess a priori for the position of the circle. We could offset that arbitrairily and say the we use (0.1,0.1) as the center of the distribution (and apply the corrections here and there to make that work), but that seems like a hack to avoid an issue more than it is a solution to the issue.
2. We have multiple plateaux in our cost function linked to the pixel size and, if we use anti-aliasing, the oversampling factor in our renderer. If two positions of the circle differ by less than the size of those pixels / oversamples, their renders will be identical, and the cost function will return the same cost.

And the issue there is really visible when you consider that at (x0,x1)=(0.0,0.0), the effective step will always be the machine precision, no matter what value I give for diff_step.

**Comment by squirion** [comment]
Personally, I feel like being able to provide a minimum absolute step would be the sound solution here, as it helps us represent correctly one more information that we have about the problem: the 'resolution' of the cost function.

**Comment by andyfaff** [comment]
For that particular problem I think there's better way of solving it, assuming that the circle is visible all the way around its circumference.
Go through the image and make a list of all the white pixel coordinates. Calculate the mean for all x values, and all y-values. That will be the centre of the circle.

Alternatively it'd probably be easier to use something like `cv2.HoughCircles` (https://theailearner.com/tag/cv2-houghcircles/).

**Comment by squirion** [comment]
Yes, I know, I was just trying to find a toy problem that exhibited the characteristics of the much more complexe problem I'm trying to solve, which has plateaux and for which we search at or around x=0

**Comment by andyfaff** [comment]
Adding a minimum absolute step to `least_squares` would have the following things to consider:

- `min_abs_step` would indicate that there should be an `abs_step` as well.
- what if you only want `min_abs_step` to apply to one parameter and not others?
- which would take preference, `abs_step` or `diff_step`?
- what happens if you want `diff_step` for some parameters and `abs_step` for others?
- does `min_abs_step` just clip the relative step?
- ... 

I can see the point of what you want, but it's also increasing complexity by a lot as well.

Another alternative if you want to pursue using least squares would be to specify your own `jac` callable. You can then specify whatever finite differencing scheme you want:

```
from scipy.optimize import approx_fprime

class my_jac:
    def __init__(self, fun, epsilon=np.sqrt(np.finfo(float).eps)):
        self.eps = epsilson
        self.fun = fun
        
    def __call__(self, xk):
        # you could create whatever absolute step (epsilon) you desire at this point
        return approx_fprime(xk, self.fun, epsilon=self.eps)
```

**Comment by squirion** [comment]
> `min_abs_step` would indicate that there should be an `abs_step` as well.

Maybe it should be called `min_diff_step`, but no, it should not imply that we need to define an `abs_step`, it's really meant to be a minimum value that we allow ` _compute_absolute_step` to return for that parameter

> what if you only want `min_abs_step` to apply to one parameter and not others?

To specify it per parameter, pass a vector, similar to how bounds are handled. A value of 0 would default to the machine precision.

> which would take preference, abs_step or diff_step?
> what happens if you want diff_step for some parameters and abs_step for others?

As mentionned above, there would be no `abs_step`, we only provide the lowest value that ` _compute_absolute_step` can return for that parameter

> does min_abs_step just clip the relative step?

Yes it would.

I think a lot of the complexity is removed when you consider my request is not about mixing relative and absolute values, but I agree that it is not a trivial change either. Still, I feel like it is something fundamental that is missing from the current implementation considering most real world problems can have a fair amount of noise and will suffer from similar limitations around x=0. In my case, I'm mostly limited by plateaux, which is a more specific issue, I can agree on that though.

So, yeah, that's pretty much the best case I can make for why I think allowing to specify a minimum step is the way to go, but if the community doesn't see the value there I can't force anyone either, and yes, specifying a `jac` would be a good workaround.

**Comment by dschmitz89** [comment]
Slightly offtopic but related to your problems: have you considered using a derivative-free optimizer? For me, `optimize(method='powell')` gave for example good results for noisy problems.

**Comment by squirion** [comment]
Thanks for the tip @dschmitz89 !
I am currently translating / refactoring Matlab code into Python, so I prefer to stick closer to the original code at the moment, but I'll definitely check it out eventually.

**Comment by andyfaff** [comment]
An alternative is to use `leastsq`, where specification of a `epsfcn` would allow you to have full control over the step size in the finite differencing.

**Comment by squirion** [comment]
Thanks @andyfaff , interesting option, but I see that this does not support bounds on the parameters, so that won't work in my case.
